### ISSUE TITLE
reserve sections of flag space

### ISSUE THREAD

**Issue by petrelharp** [tskit-dev/msprime#490]
As discussed in #484 we should reserve a section of flag space (for all flag columns in tables) for applications to use.

Proposal: we have 32 bits; let's leave the top 8 (24 and above) for applications. We do this by putting comments in the relevant bits of the code (`util.h`).

It was also wondered if we should increase flags to 64 bits. I propose not doing this, since it'd increase file size not insubstantially.

**Comment by molpopgen** [comment]
Currently, only NodeTable uses flags, right?

Keeping the flag to 32 bits is helpful for a bunch of reasons beyond the file size.  Masks for 64 bit integers are prone to platform/compiler-specific idiosyncrasies.

If we're thinking that there's enough odd biology that you start wanting lots of bits, isn't a simpler solution an array of the relevant biology?  Its length would == len(nodes) and it would simplify in linear time given the idmap returned by simplification.  Done right, the memory management would be a non-issue. At the end of a sim, this array could all be reformatted into nice Individual and/or Node metadata as needed.  This is how I'd be tempted to do things, and it has the plus side of no bit masks.

**Comment by petrelharp** [comment]
> Currently, only NodeTable uses flags, right?

yes, although the `flags` column is there for individuals, also.

**Comment by jeromekelleher** [comment]
As per [this comment](https://github.com/tskit-dev/msprime/issues/482#issuecomment-393896776) it's less clear to me know how useful this is. If flags are for communicating with ``tskit``, then by definition we shouldn't reserve part of the space for application use. Applications should use ``metadata`` to encode information that users need to know, but ``tskit`` doesn't care about.

What do we think? I like this distinction because I think we're going to repeatedly find ourselves going down rabbit holes trying to define what should be in flags and what shouldn't. With this it's clear: if ``tskit`` uses it, it's a flag. If not it isn't.

**Comment by petrelharp** [comment]
Agree. As you said previously, it doesn't make sense to make standards for something we aren't actually using.
